include/boost/geometry/algorithms/for_each.hpp

```diff
@@ -94,31 +95,88 @@ struct fe_range_per_point
 };
 
 
+template <typename Point, typename Reference, bool IsConst>
+struct fe_range_per_segment_call_f;
+
+template <typename Point>
+struct fe_range_per_segment_call_f<Point, Point&, true>
+{
+    template <typename Iterator, typename Functor>
+    static inline void apply(Iterator it0, Iterator it1, Functor& f)
+    {
+        model::referring_segment<Point const> s(*it0, *it1);
+        f(s);
+    }
+};
+
+template <typename Point>
+struct fe_range_per_segment_call_f<Point, Point&, false>
+{
+    template <typename Iterator, typename Functor>
+    static inline void apply(Iterator it0, Iterator it1, Functor& f)
+    {
+        model::referring_segment<Point> s(*it0, *it1);
+        f(s);
+    }
+};
+
+template <typename Point, typename Reference>
+struct fe_range_per_segment_call_f<Point, Reference, true>
+{
+    template <typename Iterator, typename Functor>
+    static inline void apply(Iterator it0, Iterator it1, Functor& f)
+    {
+        Point const& p0 = *it0;
+        Point const& p1 = *it1;
+        model::referring_segment<Point const> s(p0, p1);
+        f(s);
+    }
+};
+
+template <typename Point, typename Reference>
+struct fe_range_per_segment_call_f<Point, Reference, false>
+{
+    template <typename Iterator, typename Functor>
+    static inline void apply(Iterator it0, Iterator it1, Functor& f)
+    {
+        Point p0 = *it0;
+        Point p1 = *it1;
+        model::referring_segment<Point> s(p0, p1);
+        f(s);
+        *it0 = p0;
+        *it1 = p1;
+    }
+};
+
+
 template <closure_selector Closure>
 struct fe_range_per_segment_with_closure
 {
     template <typename Range, typename Functor>
     static inline void apply(Range& range, Functor& f)
     {
-        typedef typename add_const_if_c
-            <
-                is_const<Range>::value,
-                typename point_type<Range>::type
-            >::type point_type;
-
         typedef typename boost::range_iterator<Range>::type iterator_type;
 
         iterator_type it = boost::begin(range);
-        if (it == boost::end(range))
+        iterator_type end = boost::end(range);
+        if (it == end)
         {
             return;
         }
 
         iterator_type previous = it++;
-        while(it != boost::end(range))
+        while(it != end)
         {
-            model::referring_segment<point_type> s(*previous, *it);
-            f(s);
+            fe_range_per_segment_call_f
+                <
+                    typename add_const_if_c
+                        <
+                            is_const<Range>::value,
+                            typename point_type<Range>::type
+                        >::type,
+                    typename boost::range_reference<Range>::type,
+                    is_const<Range>::value
+                >::apply(previous, it, f);
             previous = it++;
```

Comment:
This is quite a complex solution to add a const, or not...
Isn't a simpler type possible? 
And in C++11/14 (with auto) maybe?

Just questions, I'm fine anyway.

Reply:
I can't think of anything else besides encapsulating it inside `fe_range_per_segment_call_f` so at least it wouldn't be required to write it twice, here and below.

At least as long as we want to preserve the current behavior. This is an implementation of `for_each_segment` and calling this for mutable ranges probably has no sense because there are no separate segments in them so if a user modified a point of a mutable segment then another segment would be modified indirectly. So a simple and reasonable fix would be to remove all of this `is_const` logic at all and always require the Range to be `const`. In this case also the `fe_range_per_segment_call_f` would be not needed at all.

Side info: `for_each_segment ` is not used anywhere in the library. At least for now. We use `segment_iterator` instead. But when we move to C++14 I think I'd prefer to use it (or rather implement a similar algorithm `for_each_segment_if` allowing to break the iteration) together with generic lambdas instead of `segment_iterator` because it's a lot lighter.

@mloskot do you have some other idea?

Reply:
Thanks for your explanation. Indeed - using `for_each_segment` doesn't really make sense. But removing it is a breaking change.

Reply:
Sorry, no better ideas. If nothing requires passing non-const, then narrowing the type qualification down to single possibility is certainly a good idea, and a  simplification.

Reply:
@barendgehrels ok, I simplified it preserving backward compatibility.

Reply:
Cool! Thank you!